modules/020-deckhouse/hooks/internal/updater/updater.go

```diff
@@ -0,0 +1,654 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/utils/pointer"
+
+	"github.com/deckhouse/deckhouse/go_lib/dependency/requirements"
+	"github.com/deckhouse/deckhouse/go_lib/hooks/update"
+	"github.com/deckhouse/deckhouse/modules/020-deckhouse/hooks/internal/apis/v1alpha1"
+)
+
+const (
+	metricReleasesGroup = "d8_releases"
+)
+
+type DeckhouseUpdater struct {
+	now          time.Time
+	inManualMode bool
+
+	// probably we have to change to interfaces but later
+	input *go_hook.HookInput
+
+	// don't modify releases order, logic is based on this sorted slice
+	releases                   []DeckhouseRelease
+	totalPendingManualReleases int
+
+	predictedReleaseIndex       int
+	skippedPatchesIndexes       []int
+	currentDeployedReleaseIndex int
+	forcedReleaseIndex          int
+	predictedReleaseIsPatch     *bool
+
+	deckhousePodIsReady      bool
+	deckhouseIsBootstrapping bool
+
+	releaseData        DeckhouseReleaseData
+	notificationConfig *NotificationConfig
+}
+
+func NewDeckhouseUpdater(input *go_hook.HookInput, mode string, data DeckhouseReleaseData, podIsReady, isBootstrapping bool) *DeckhouseUpdater {
+	nConfig := ParseNotificationConfigFromValues(input)
+	now := time.Now().UTC()
+	if os.Getenv("D8_IS_TESTS_ENVIRONMENT") != "" {
+		now = time.Date(2021, 01, 01, 13, 30, 00, 00, time.UTC)
+	}
+	return &DeckhouseUpdater{
+		now:                         now,
+		inManualMode:                mode == "Manual",
+		input:                       input,
+		predictedReleaseIndex:       -1,
+		currentDeployedReleaseIndex: -1,
+		forcedReleaseIndex:          -1,
+		skippedPatchesIndexes:       make([]int, 0),
+		deckhousePodIsReady:         podIsReady,
+		deckhouseIsBootstrapping:    isBootstrapping,
+		releaseData:                 data,
+		notificationConfig:          nConfig,
+	}
+}
+
+// for patch we check less conditions, then for minor release
+// - Canary settings
+func (du *DeckhouseUpdater) checkPatchReleaseConditions(predictedRelease *DeckhouseRelease) bool {
+	// check: canary settings
+	if predictedRelease.ApplyAfter != nil {
+		if du.now.Before(*predictedRelease.ApplyAfter) {
+			du.input.LogEntry.Infof("Release %s is postponed by canary process. Waiting", predictedRelease.Name)
+			du.updateStatus(predictedRelease, fmt.Sprintf("Release is postponed until: %s", predictedRelease.ApplyAfter.Format(time.RFC822)), v1alpha1.PhasePending)
+			return false
+		}
+	}
+
+	return true
+}
+
+func (du *DeckhouseUpdater) checkReleaseNotification(predictedRelease *DeckhouseRelease, updateWindows update.Windows) bool {
+	if du.inManualMode || du.releaseData.Notified {
+		return true
+	}
+
+	var applyTimeChanged bool
+	predictedReleaseApplyTime := du.predictedReleaseApplyTime(predictedRelease)
+	if du.notificationConfig.MinimalNotificationTime.Duration > 0 {
+		minApplyTime := du.now.Add(du.notificationConfig.MinimalNotificationTime.Duration)
+		if minApplyTime.Before(predictedReleaseApplyTime) {
+			minApplyTime = predictedReleaseApplyTime
```

Comment:
Can you add (or show me) test for this code. I broke my mind when i read this

Reply:
"Notification: release applyAfter time is after notification period" test is about that